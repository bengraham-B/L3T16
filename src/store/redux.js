import { createSlice } from '@reduxjs/toolkit'

export const reduxSlice = createSlice({
    name: "redux",

    initialState: {
        userAuthStatus: false,
        refreshCountValue: 0,
        error: null
    },

    reducers: {
        authStatus: (state) => {
            if(localStorage.getItem('goose-reloaded-user')){
                return {
                    ...state,
                    userAuthStatus: true
                }
            }

            if(!localStorage.getItem('goose-reloaded-user')){
                return {
                    ...state,
                    userAuthStatus: false
                }

            }
        },

        logoutRedux: (state) => {
            window.location.assign("/auth/login") //^ when the user logs out their will be sent to the logout page
            localStorage.removeItem("goose-reloaded-user") //^ Removes JWT from local storage

            //^ Updates the state 
            return {
                ...state,
                userAuthStatus: false
            }

        },

        loginRedux: async (state, props) => {

            try {
                const response = await fetch("http://localhost:8001/api/user/login", {
                    method: "POST",
                    body: JSON.stringify({
                        email: props.payload.email,
                        password: props.payload.password
                    }),
                    headers:{
                        "Content-Type": "application/json"
                    }
                })

                const data = await response.json()

                if(data.error){
                    console.log(data.error)
                }

                else{
                    localStorage.setItem("goose-reloaded-user", JSON.stringify(data))
                    window.location.assign("/")

                    return {
                        ...state,
                        userAuthStatus: true
                    }
                }
            }

            catch(error){
                state.error = "Could not connect to the server"
            }


        },
           
        //^ This function is responsible for signup the user
        signupRedux: async (state, props) => {

            try {
                //^ sending data to the server, where the user will be authenticated.
                const response = await fetch("http://localhost:8001/api/user/signup", {
                    method: "POST",
                    body: JSON.stringify({
                        email: props.payload.email,
                        password: props.payload.password
                    }),
                    headers: {
                        "COntent-Type": "application/json"
                    }
                })

                const data = await response.json()

                //^ If the response is okay, the JWT will be saved to local stroage and the user will be sent to the home page.
                if(response.ok){
                    localStorage.setItem("goose-reloaded-user", JSON.stringify(data))
                    window.location.assign("/")

                    return {
                        ...state,
                        authStatus: true
                    }
                }
            }

            //^ If the React application is unable to connect to the server this error will be thrown.
            catch(error){
                state.error = "Could not connect to server"
            }

        },

        refreshCount: (state) => {

            state.refreshCountValue = state.refreshCountValue + 1

            console.log("REDUX", state.refreshCountValue )

            // return {
            //     ...state,
            //     refreshCountValue: state.refreshCountValue + 1
            // }
        }

    }
})

export const { authStatus, logoutRedux, loginRedux, signupRedux, refreshCount } = reduxSlice.actions;
export default reduxSlice.reducer