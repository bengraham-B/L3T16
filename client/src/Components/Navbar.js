import React, {useState, useEffect} from 'react'
import { Link } from 'react-router-dom'
import {useDispatch, useSelector} from 'react-redux'

import { authStatus, logoutRedux } from '../store/redux'

export default function Navbar() {
	const user = useSelector((state) => state.redux.userAuthStatus)
	console.log(user)

	const [userName, setUserName] = useState()

	const dispatch = useDispatch()

	const handleLogout = () => {
		dispatch(logoutRedux())
		dispatch(authStatus())
	}

	dispatch(authStatus())
	console.log(user)

	

	useEffect(() => {

		if(localStorage.getItem("goose-reloaded-user")){
			const object = JSON.parse(localStorage.getItem("goose-reloaded-user"))
			setUserName(object.email)
		} else {
			setUserName("")
		}

	}, [])

	return (
		<div id="Navbar">
			<div className="home-container">
				<Link to="/">
					<h2>Home</h2>
				</Link>
			</div>

			<div className="title-container">
				<h1>GOOSE_RELOADED</h1>
			</div>

			<div className="auth-container">
				{user ? 
					<div className='button-conatiner'>
						<h3>{userName}</h3> 
						<button onClick={handleLogout}>Logout</button>
					</div> 
					:
					<div className='button-conatiner'>
						<button className='unauthed-user'>
							<Link to="/auth/login">Login</Link>
						</button>
						<button className='unauthed-user'>
							<Link to="/auth/signup">Signup</Link>
						</button>

					</div>

					
					}
				
			</div>
		</div>
	)
}
