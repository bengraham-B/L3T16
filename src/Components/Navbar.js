import React, {useState, useEffect} from 'react'
import { Link } from 'react-router-dom'
import {useDispatch, useSelector} from 'react-redux'

import { authStatus, logoutRedux } from '../store/redux'

export default function Navbar() {
	const user = useSelector((state) => state.redux.userAuthStatus)
	console.log(user)

	const [userName, setUserName] = useState()
	const [admin, setAdmin] = useState()

	const dispatch = useDispatch()

	const handleLogout = () => {
		dispatch(logoutRedux())
		dispatch(authStatus())
	}

	dispatch(authStatus())
	console.log(user)

	

	useEffect(() => {

		if(localStorage.getItem("goose-reloaded-user")){
			const object = JSON.parse(localStorage.getItem("goose-reloaded-user"))
			setUserName(object.email)

			//^ checking if the user is an admin
			if(object.admin){
				setAdmin("Admin")
			} else {
				setAdmin("")
			}
			
		} else {
			setUserName("")
		}

	}, [])

	return (
		<div id="Navbar">
			<div className="home-container">
				<Link to="/">
					<h2>Home</h2>
				</Link>
			</div>

			<div className="title-container">
				<h1>GOOSE_RELOADED</h1>
				{/* If the user is an admin will display, it will display admin under the title. */}
				<h3>{admin}</h3>
			</div>

			<div className="auth-container">
				{user ? 
					<div className='button-conatiner'>
						<h3>{userName} </h3> 
						<button onClick={handleLogout}>Logout</button>
					</div> 
					:
					<div className='button-conatiner'>
						<button className='unauthed-user'>
							<Link to="/auth/login">Login</Link>
						</button>
						<button className='unauthed-user'>
							<Link to="/auth/signup">Signup</Link>
						</button>

					</div>

					
					}
				
			</div>
		</div>
	)
}
