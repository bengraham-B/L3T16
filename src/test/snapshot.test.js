//^ Creating a Snapshot test for the Navbar component
import React from 'react';
import { render } from '@testing-library/react';
import Navbar from '../Components/Navbar'
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import store from '../store/store'


test('renders Navbar correctly', () => {
    const { container } = render(
        <Provider store={store}>
            <BrowserRouter basename="/">
                <Navbar />
            </BrowserRouter>
        </Provider>
    
    );
    expect(container).toMatchSnapshot();
  });
  