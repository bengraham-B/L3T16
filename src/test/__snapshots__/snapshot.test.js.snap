// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Navbar correctly 1`] = `
<div>
  <div
    id="Navbar"
  >
    <div
      class="home-container"
    >
      <a
        href="/"
      >
        <h2>
          Home
        </h2>
      </a>
    </div>
    <div
      class="title-container"
    >
      <h1>
        GOOSE_RELOADED
      </h1>
      <h3 />
    </div>
    <div
      class="auth-container"
    >
      <div
        class="button-conatiner"
      >
        <button
          class="unauthed-user"
        >
          <a
            href="/auth/login"
          >
            Login
          </a>
        </button>
        <button
          class="unauthed-user"
        >
          <a
            href="/auth/signup"
          >
            Signup
          </a>
        </button>
      </div>
    </div>
  </div>
</div>
`;
